lib/rspec/matchers/expecteds_for_multiple_diffs.rb

```diff
@@ -0,0 +1,118 @@
+module RSpec
+  module Matchers
+    # @api private
+    # Handles list of expected values when there is a need to render
+    # multiple diffs. Also can handle one value.
+    class ExpectedsForMultipleDiffs
+      # @api private
+      # Wraps provided expected value in instance of
+      # ExpectedForMultipleDiffs. If provided value is already an
+      # ExpectedForMultipleDiffs then it just returns it.
+      # @param [Any] expected value to be wrapped
+      # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+      def self.from(expected)
+        return expected if self === expected
+        new.add(expected)
+      end
+
+      # @api private
+      # Appends expected value to itself. Returns itself for possible
+      # chaining.
+      # @param [Any] expected value to append to itself
+      # @param [Hash] options optional keyword arguments, for example for_matcher
+      # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+      def add(expected, options={})
+        for_matcher = options.fetch(:for_matcher, NoMatcher.new)
+        expected_list.concat(expected_from(expected, ForMatcher.from(for_matcher)))
+        self
+      end
+
+      # @api private
+      # Returns message with diff(s) appended for provided differ
+      # factory and actual value if there are any
+      # @param [String] message original failure message
+      # @param [Proc] differ_factory brand new differ factory method (or proc, block)
+      # @param [Any] actual value
+      # @return [String]
+      def message_with_diff(message, differ_factory, actual)
+        diff = diffs(differ_factory, actual)
+        message = "#{message}\n#{diff}" unless diff.empty?
+        message
+      end
+
+    protected
+
+      # @api private
+      # Returns list of expected values.
+      # @return [Array<RSpec::Matchers::ExpectedsForMultipleDiffs::Expected>]
+      def expected_list
+        @_expected_list ||= []
+      end
+
+    private
+
+      def diffs(differ_factory, actual)
+        expected_list.map do |expected|
+          expected.diffs(differ_factory, actual)
+        end.inject([], :+).join("\n")
```

Comment:
This method iterates over the list twice -- once for `map`, once for `inject`.  I suspect it will perform better to iterate only once.  Can you use `flat_map` instead and only iterate once?

Note that 1.8.7 lacks `Enumerable#flat_map`; in rspec-core we defined our own wrapper that deals with that:

https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/flat_map.rb

(See also rspec/rspec-core#1854 -- there's a mistake in that file). While you can't directly use rspec-core from rspec-expectations for that, you could do something like:

``` ruby
module RSpec::Expectations
  if defined?(RSpec::Core::FlatMap
    FlatMap = RSpec::Core::FlatMap
  else
    # define it inline using an identical definition to what rspec-core has
  end
end
```

It's a bit non-DRY but I'm not really concerned since `flat_map` isn't a piece of knowledge that's likely to evolve.

BTW, we do have benchmarks in rspec-core showing that `flat_map` is faster than `inject` -- see rspec/rspec-core#1836.